server/viame_tasks/tasks.py

```diff
@@ -6,72 +7,118 @@
-            "-s detector_writer:file_name={}".format(output_path),
+            "kwiver runner",
+            f"-p {pipeline_path}",
+            f"-s input:video_filename={image_list_file}",
+            f"-s detector_writer:file_name={detector_output_path}",
+            f"-s track_writer:file_name={track_output_path}",
         ]
+    else:
+        raise ValueError('Unknown input type: {}'.format(input_type))
 
-    process = Popen(" ".join(command), stderr=PIPE, stdout=PIPE, shell=True)
+    cmd = " ".join(command)
+    print('Running command:', cmd)
+    process = Popen(cmd, stderr=PIPE, stdout=PIPE, shell=True, executable='/bin/bash')
     stdout, stderr = process.communicate()
-    output = str(stdout) + "\n" + str(stderr)
-    self.job_manager.write(output)
-    return output_path
+    if process.returncode > 0:
+        raise RuntimeError('Pipeline exited with nonzero status code {}: {}'.format(process.returncode, str(stderr)))
+    else:
+        self.job_manager.write(str(stdout) + "\n" + str(stderr))
+
+    # Figure out which of track_output_path, detector_output_path to return
+    # Some pipelines don't produce track output, so fallback to returning detector path
+    # Both files WILL exist, but if a file hasn't been written to, it will be 0 bytes
+    if os.path.getsize(track_output_path) > 0:
+        os.remove(detector_output_path)
+        return track_output_path
+    else:
+        os.remove(track_output_path)
+        return detector_output_path
 
 
 @app.task(bind=True)
 def convert_video(self, path, folderId, token, auxiliaryFolderId):
     self.girder_client.token = token
+
+    # Delete is true, so the tempfile is deleted when the block closes.
+    # We are only using this to get a name, and recreating it below.
     with tempfile.NamedTemporaryFile(suffix=".mp4", delete=True) as temp:
         output_path = temp.name
 
+    # Extract metadata
     file_name = os.path.join(path, os.listdir(path)[0])
-    process = Popen(["ffmpeg", "-i", file_name], stderr=PIPE, stdout=PIPE)
-    stdout, stderr = process.communicate()
-    output = str(stdout) + "\n" + str(stderr)
-    fps_string = [i for i in output.split(",") if "fps" in i][0]
-    fps_value = [int(s) for s in fps_string.split() if s.isdigit()][0]
-    self.girder_client.addMetadataToFolder(folderId, {"fps": fps_value})
+    command = [
+        "ffprobe",
+        "-print_format", "json",
+        "-v", "quiet",
+        "-show_format",
+        "-show_streams",
+        file_name,
+    ]
+    process = Popen(command, stderr=PIPE, stdout=PIPE)
+    stdout = process.communicate()[0]
+    jsoninfo = json.loads(stdout.decode('utf-8'))
+    videostream = list(filter(lambda x: x["codec_type"] == "video", jsoninfo["streams"]))
+    if len(videostream) != 1:
+        raise Exception('Expected 1 video stream, found {}'.format(len(videostream)))
+    self.girder_client.addMetadataToFolder(folderId, videostream[0])
```

Comment:
Was the intention of this just for reference or was it supposed to override the fps data that is specified during upload?
If it should be added for reference should it be contained in it's own JSON object to prevent it from polluting the root of the metadata?
![image](https://user-images.githubusercontent.com/61746913/81108046-1a1dbc00-8ee6-11ea-8fef-6ae30d32cb3f.png)

Reply:
It was sort of a "we're going to all the trouble to compute this info, might as well hang onto it".

Also the FPS you get isn't usable because real FPS fractions can be nasty.

I think nesting it into it's own json object is a great suggestion.

Reply:
Done in c0bbd61